homeassistant/components/light/yeelight.py

```diff
@@ -6,157 +6,311 @@
-
-    def turn_on(self, **kwargs):
-        """Turn the specified or all lights on."""
-        if not self.is_on:
-            self._bulb.turn_on()
-
-        if ATTR_RGB_COLOR in kwargs:
-            rgb = kwargs[ATTR_RGB_COLOR]
-            self._bulb.set_rgb_color(rgb[0], rgb[1], rgb[2])
-            self._rgb = [rgb[0], rgb[1], rgb[2]]
-
-        if ATTR_COLOR_TEMP in kwargs:
-            kelvin = int(mired_to_kelvin(kwargs[ATTR_COLOR_TEMP]))
-            self._bulb.set_color_temperature(kelvin)
-            self._ct = kelvin
-
-        if ATTR_BRIGHTNESS in kwargs:
-            bright = int(kwargs[ATTR_BRIGHTNESS] * 100 / 255)
-            self._bulb.set_brightness(bright)
-            self._bright = kwargs[ATTR_BRIGHTNESS]
-
-    def turn_off(self, **kwargs):
-        """Turn the specified or all lights off."""
-        self._bulb.turn_off()
-
-    def update(self):
-        """Synchronize state with bulb."""
-        self._bulb.refresh_property()
-
-        # Update power state
-        self._state = self._bulb.get_property(self._bulb.PROPERTY_NAME_POWER)
-
-        # Update Brightness value
-        bright_percent = self._bulb.get_property(
-            self._bulb.PROPERTY_NAME_BRIGHTNESS)
-        bright = int(bright_percent) * 255 / 100
-        # Handle 0
-        if int(bright) == 0:
-            self._bright = 1
+    def _bulb(self) -> object:
+        import yeelight
+        if self.__bulb is None:
+            try:
+                self.__bulb = yeelight.Bulb(self._ipaddr)
+                self.__bulb.get_properties()  # force init for type
+
+                btype = self.__bulb.bulb_type
+                if btype == yeelight.BulbType.Color:
+                    self._supported_features |= SUPPORT_YEELIGHT_RGB
+                self._available = True
+            except (yeelight.BulbException, socket.error, OSError) as ex:
+                self._available = False
+                _LOGGER.error("Failed to connect to bulb %s, %s: %s",
+                              self._ipaddr, self._name, ex)
+
+        return self.__bulb
+
+    def set_music_mode(self, mode) -> None:
+        """Set the music mode on or off."""
+        if mode:
+            self._bulb.start_music()
         else:
-            self._bright = int(bright)
-
-        # Update RGB Value
-        raw_rgb = int(
-            self._bulb.get_property(self._bulb.PROPERTY_NAME_RGB_COLOR))
-        red = int(raw_rgb / 65536)
-        green = int((raw_rgb - (red * 65536)) / 256)
-        blue = raw_rgb - (red * 65536) - (green * 256)
-        self._rgb = [red, green, blue]
-
-        # Update CT value
-        self._ct = int(self._bulb.get_property(
-            self._bulb.PROPERTY_NAME_COLOR_TEMPERATURE))
+            self._bulb.stop_music()
+
+    def update(self) -> None:
+        """Update properties from the bulb."""
+        import yeelight
+        try:
+            self._bulb.get_properties()
+
+            is_on = self._properties.get("power", STATE_OFF)
```

Comment:
Don't use `STATE_OFF` here, since that's a home assistant constant, and it's getting compared to a device library property. It's value happens to be the same, but as convention we treat the "STATE_*" constants as only valid to compare against home assistant states.

You could also shorten these lines to something like this:
`self._is_on = self._properties.get("power") == "on"`. There's no need to use an if and then only assign a boolean.